### PR TITLE
Add RepoClient as wrapper that auto-detects and supports both git & hg

### PR DESCRIPTION
**Pull request by eemeli**
~This builds on top of #5, so at least currently [these](https://github.com/mozilla/fluent-migrate/pull/6/files/2defc24db8e2e0a4b422b70da710a3de4b4a2eb3..90362e1179761b7c1ca7756bfbc406b12bcd1ff6) are the only proper changes here.~

A new `RepoClient` class is added, with a minimal set of methods as required by the rest of the library for repository operations. A Mercurial repo is detected from the presence of a `.hg` directory, and for Git we look for `.git` and check that `git rev-parse --is-inside-work-tree` returns `true`.

The `hglib` stuff should work just as before; for git we don't need any wrapper library as the command line is plenty. For the blame I did need to write a minimal parser, but that was only about 10 lines of code.

The `Blame().handleFile()` function signature changed a bit through all this, but AFAIK that has no external users.

### CODE REVIEW

**Comment by flodolo** [comment]
I wonder if we could have the "apply black formatting" as a completely separate PRs? It makes reviewing unnecessarily hard, even with separate commits.

I'd be OK with you landing that change directly without review, since that's the standard we use all over the place.

**Comment by eemeli** [comment]
> I'd be OK with you landing that change directly without review, since that's the standard we use all over the place.

Done, and rebased also this PR.

**Comment by eemeli** [comment]
Rebased as #5 was merged. Also exposed `git()` as a utility function separately from `RepoClient` + added `head()` and `log()` methods to support usage from `test_fluent_migrations` in m-c.

File: fluent/migrate/repo_client.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    """
+    Wrapper for calling command-line git in the `root` directory.
+    Raises an exception on any error, including a non-0 return code.
+    Returns the command's stdout as a string.
+    """
+    git = ["git"]
+    git.extend(args)
+    proc = run(git, capture_output=True, cwd=root, encoding="utf-8")
+    if proc.returncode != 0:
+        raise Exception(proc.stderr or f"git {args[0]} failed")
```
Comment (flodolo) [inline fluent/migrate/repo_client.py]: This would log only the first parameter (e.g. `add`, `commit`)? If so, should we log the entire command instead?

File: fluent/migrate/repo_client.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+                    self.root,
+                    "log",
+                    "--pretty=format:%s",
+                    f"{from_commit}..{to_commit}",
+                )
+                .strip()
+                .splitlines()
+            )
+
+    def _git(self, *args: str):
```
Comment (flodolo) [inline fluent/migrate/repo_client.py]: Why is this needed? 🤔

File: fluent/migrate/repo_client.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+import json
+from subprocess import run
+
+from os.path import isdir, join
+
+import hglib
+
+
+def git(root: str, *args: str) -> str:
```
Comment (flodolo) [inline fluent/migrate/repo_client.py:12]: Have you considered (and discarded) the idea of using a library similar to hglib for git?

**Review by bcolsson — APPROVED** [review]
Reviewed this and nothing stands out to me so approving. Next time there's a migration ready I'll attempt a test with this.

**Review by flodolo — APPROVED** [review]

File: fluent/migrate/repo_client.py (hunk 3 above)
Comment (eemeli) [reply fluent/migrate/repo_client.py:12]: Yes. This way we avoid a layer of indirection in needing to refer to an external definition for what `git()` does, and have instead this minimal six-line function that's sufficient for our purposes.

File: fluent/migrate/repo_client.py (hunk 1 above)
Comment (eemeli) [reply fluent/migrate/repo_client.py]: I would be surprised if the `or` condition here ever failed for the first part, so the issue is pretty academic. When failing, `git` is really rather good about telling you how/why.

But sure, we could to this:
```suggestion
        raise Exception(proc.stderr or f"git command failed: {args}")
```

Please accept this suggestion if you'd prefer that.

File: fluent/migrate/repo_client.py (hunk 2 above)
Comment (eemeli) [reply fluent/migrate/repo_client.py]: Good catch, it isn't. This is a leftover from refactoring.

File: fluent/migrate/repo_client.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+                    "log",
+                    "--pretty=format:%s",
+                    f"{from_commit}..{to_commit}",
+                )
+                .strip()
+                .splitlines()
+            )
+
+    def _git(self, *args: str):
+        return git(self.root, *args)
```
Comment (eemeli) [inline fluent/migrate/repo_client.py]:
```suggestion
```
